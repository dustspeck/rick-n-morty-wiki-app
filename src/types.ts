interface IPageInfo {
  count: number;
  pages: number;
  next: string | null;
  prev: string | null;
}
export interface IPage {
  info: IPageInfo;
  results: any[];
}

interface IOrigin {
  name: string;
  url: string;
}

interface ILocation {
  name: string;
  url: string;
}

export type IStatus = 'Alive' | 'Dead' | 'unknown';
export type IGender = 'Female' | 'Male' | 'Genderless' | 'unknown';

export interface ICharacter {
  id: number;
  name: string;
  status: IStatus;
  species: string;
  type: string;
  gender: IGender;
  origin: IOrigin;
  location: ILocation;
  image: string;
  episode: string[];
  url: string;
  created: string;
}

export interface ICharactersPage extends IPage {
  results: ICharacter[];
}
