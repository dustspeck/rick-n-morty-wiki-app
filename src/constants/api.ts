export const BASE_URL = 'https://rickandmortyapi.com/api';
export const CHARACTERS_ROUTE = `${BASE_URL}/character`;
