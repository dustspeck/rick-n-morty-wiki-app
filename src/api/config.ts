import axios from 'axios';
import {BASE_URL} from '../constants/api';

export const apiClient = axios.create({baseURL: BASE_URL});
