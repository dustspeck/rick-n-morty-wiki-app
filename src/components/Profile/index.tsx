import {View, Text, StyleSheet, ScrollView} from 'react-native';
import React from 'react';
import SceneBuilder from '../common/SceneBuilder';
import {BLACK, GREY} from '../../constants/colors';
import {ROOT_PADDING} from '../../constants/styles';
import ProfileImage from './ProfileImage';
import PropSection from './PropSection';
import WhereaboutsSection from './WhereaboutsSection';
import ChaptersSection from './ChaptersSection';
import {IProfileProps} from '../../types';

const Profile = (props: IProfileProps) => {
  const character = props.route.params.data;

  return (
    <SceneBuilder showTopBar={true}>
      <ScrollView
        centerContent
        showsVerticalScrollIndicator={false}
        style={styles.container}>
        <View>
          {/* Profile Image */}
          <ProfileImage character={character} />
          {/* Character Name */}
          <Text numberOfLines={2} style={styles.name}>
            {character.name}
          </Text>
          {/* Basic Properties */}
          <PropSection character={character} />
          {/* Advanced Properties */}
          <WhereaboutsSection
            origin={character.origin}
            location={character.location}
          />
          {/* Advanced Properties */}
          <ChaptersSection episodes={character.episode} />
          {/* End Note */}
          <View style={{height: 200}}>
            <Text style={styles.footerText}>That's all we know!</Text>
          </View>
        </View>
      </ScrollView>
    </SceneBuilder>
  );
};

export default Profile;

const styles = StyleSheet.create({
  container: {marginHorizontal: -ROOT_PADDING},
  footerText: {
    color: `${GREY}55`,
    fontSize: 14,
    marginTop: 25,
    textAlign: 'center',
  },
  name: {
    color: BLACK,
    fontSize: 28,
    width: '100%',
    padding: 12,
    textAlign: 'center',
    fontWeight: 'bold',
  },
});
