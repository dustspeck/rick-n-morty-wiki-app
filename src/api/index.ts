import {AxiosResponse} from 'axios';
import {CHARACTERS_ROUTE} from '../constants/api';
import {ICharactersPage, IEpisode, ILocationAll} from '../types';
import {apiClient} from './config';

/**
 * @description Fetches characters without any pagination.
 * Used for initial characters load.
 * @returns {Promise<ICharactersPage | null>} Returns the first characters page.
 * Returns `null` on errors.
 */
export const getCharacters = async (): Promise<ICharactersPage | null> => {
  try {
    const response: AxiosResponse = await apiClient.get(CHARACTERS_ROUTE);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * @description Fetches characters from the given page url.
 * @param {string} url URL to be fetched from.
 * @returns {Promise<ICharactersPage | null>} Returns the nth characters page.
 * * Returns `null` on errors.
 */
export const getCharactersByURL = async (
  url: string,
): Promise<ICharactersPage | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * Fetches location from the given url.
 * @param { string} url URL to be fetched from.
 * @returns {Promise<ILocationAll | null>} Returns the nth location page.
 * * Returns `null` on errors.
 */
export const getLocationByURL = async (
  url: string,
): Promise<ILocationAll | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ILocationAll = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * Fetches episode from the given url.
 * @param { string} url URL to be fetched from.
 * @returns {Promise<IEpisode | null>} Returns the nth episode.
 * * Returns `null` on errors.
 */
export const getEpisodeByURL = async (
  url: string,
): Promise<IEpisode | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: IEpisode = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};
