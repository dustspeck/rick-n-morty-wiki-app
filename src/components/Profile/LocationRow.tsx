import {ActivityIndicator, StyleSheet, Text, View} from 'react-native';
import React, {useState} from 'react';
import {ILocationAll, ILocation} from '../../types';
import PropRow from './PropRow';
import {getLocationByURL} from '../../services/api/api';
import {CARD_BG, GREY} from '../../constants/colors';
import {BORDER_RADIUS} from '../../constants/styles';

/**
 * Component for displaying basic `location` information of a character.
 * This is `expandable` and displays all location properties after fetching.
 */
interface LocationRowProps {
  type: string;
  location: ILocation;
}
const LocationRow = ({location, type}: LocationRowProps) => {
  // stores the fetched location data
  const [fetchedLocationData, setFetchedLocationData] =
    useState<ILocationAll | null>(null);

  // stores if the location fetch was successful
  const [fetchError, setFetchError] = useState<boolean>(false);

  const handleRowPress = async () => {
    try {
      if (fetchedLocationData) return; // data already fetched
      const locationData = await getLocationByURL(location.url);
      if (!locationData) throw 'Error fetching';
      setFetchedLocationData(locationData);
    } catch (error) {
      console.log(error);
      setFetchError(true);
    }
  };

  return (
    <PropRow
      type={type}
      value={location.name}
      // row must not be expandable for 'unknown' locations (ie. empty location url)
      // erroneous fetching will also make the row not expandable
      expandable={location.url && !fetchError ? true : false}
      onPress={handleRowPress}
      ExpandedChild={
        <LocationExpandedCard fetchedLocationData={fetchedLocationData} />
      }
    />
  );
};

export default LocationRow;

/**
 * Component used for displaying all the properties of the location.
 * This is passed to the `PropRow` as a `ExpandedChild`.
 */
interface LocationExpandedCardProps {
  fetchedLocationData: ILocationAll | null;
}
const LocationExpandedCard = ({
  fetchedLocationData,
}: LocationExpandedCardProps) => {
  if (!fetchedLocationData)
    return (
      <View style={styles.expandedContainer}>
        <ActivityIndicator color={GREY} />
      </View>
    );
  return (
    <View style={styles.expandedContainer}>
      <Text style={styles.expandedText}>
        {'Name: '}
        <Text style={styles.underlinedText}>{fetchedLocationData.name}</Text>
      </Text>
      <Text style={styles.expandedText}>
        {'Dimension: '}
        <Text style={styles.underlinedText}>
          {fetchedLocationData.dimension}
        </Text>
      </Text>
      <Text style={styles.expandedText}>
        {'# of residents: '}
        <Text style={styles.underlinedText}>
          {fetchedLocationData.residents.length}
        </Text>
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  expandedContainer: {
    backgroundColor: `${CARD_BG}55`,
    borderRadius: BORDER_RADIUS,
    height: 80,
    padding: BORDER_RADIUS,
    marginHorizontal: '10%',
    marginBottom: 20,
    overflow: 'hidden',
  },
  expandedText: {
    color: GREY,
  },
  underlinedText: {
    textDecorationLine: 'underline',
  },
});
