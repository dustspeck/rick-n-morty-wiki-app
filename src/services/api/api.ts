import axios, {AxiosResponse} from 'axios';
import {CHARACTERS_ROUTE} from '../../constants/api';
import {ICharactersPage, ILocationAll} from '../../types';
import {apiClient} from './config';

/**
 * @description Fetches characters without any pagination.
 * Used for initial characters load.
 * @returns {Promise<ICharactersPage | null>} Returns the first characters page.
 * Returns `null` on errors.
 */
export const getCharacters = async (): Promise<ICharactersPage | null> => {
  try {
    const response: AxiosResponse = await apiClient.get(CHARACTERS_ROUTE);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * @description Fetches characters of the given page number.
 * @param {number} page Page number to be fetched.
 * @returns {Promise<ICharactersPage | null>} Returns the nth characters page.
 * * Returns `null` on errors.
 */
export const getCharactersByPage = async (
  page: number,
): Promise<ICharactersPage | null> => {
  try {
    const response: AxiosResponse = await apiClient.get(
      `${CHARACTERS_ROUTE}/page=${page}`,
    );
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};

/**
 * @description Fetches characters from the given page url.
 * @param {string} url URL to be fetched from.
 * @returns {Promise<ICharactersPage | null>} Returns the nth characters page.
 * * Returns `null` on errors.
 */
export const getCharactersByURL = async (
  url: string,
): Promise<ICharactersPage | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ICharactersPage = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};
/**
 * Fetches location from the given url.
 * @param { string} url URL to be fetched from.
 * @returns {Promise<ILocationAll | null>} Returns the nth location page.
 * * Returns `null` on errors.
 */
export const getLocationByURL = async (
  url: string,
): Promise<ILocationAll | null> => {
  try {
    if (!url) return null;
    const response: AxiosResponse = await apiClient.get(`${url}`);
    const data: ILocationAll = response.data;
    return data;
  } catch (error) {
    console.log(error);
    return null;
  }
};
