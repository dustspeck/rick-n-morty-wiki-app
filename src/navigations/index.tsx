import {createNativeStackNavigator} from '@react-navigation/native-stack';
import Home from '../components/Home';
import Profile from '../components/Profile';
import {ICharacter} from '../types';

export type RootStackParamList = {
  Profile: {data: ICharacter};
  Home: undefined;
};

const Stack = createNativeStackNavigator<RootStackParamList>();

const MainStack = () => {
  return (
    <Stack.Navigator
      initialRouteName="Home"
      screenOptions={{headerShown: false}}>
      <Stack.Screen name="Home" component={Home} />
      <Stack.Screen name="Profile" component={Profile} />
    </Stack.Navigator>
  );
};

export default MainStack;
