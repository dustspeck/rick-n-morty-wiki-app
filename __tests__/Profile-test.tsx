import {fireEvent, render, waitFor} from '@testing-library/react-native';
import Profile from '../src/screens/Profile';
import {getEpisodeByURL, getLocationByURL} from '../src/api';

import {ICharactersPage, IEpisode, ILocationAll} from '../src/types';
import {RouteProp} from '@react-navigation/native';
import {RootStackParamList} from '../src/navigations/index';
import LocationRow from '../src/components/Profile/LocationRow';

const characterPageResponse = {
  info: {count: 0, next: null, pages: 0, prev: null},
  results: [
    {
      id: 20,
      name: 'Ants in my Eyes Johnson',
      status: 'unknown',
      species: 'Human',
      type: 'Human with ants in his eyes',
      gender: 'Male',
      origin: {
        name: 'unknown',
        url: '',
      },
      location: {
        name: 'Interdimensional Cable',
        url: 'https://rickandmortyapi.com/api/location/6',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/20.jpeg',
      episode: ['https://rickandmortyapi.com/api/episode/8'],
      url: 'https://rickandmortyapi.com/api/character/20',
      created: '2017-11-04T22:34:53.659Z',
    },
  ],
} as ICharactersPage;

const episodePageResponse = {
  id: 1,
  name: 'Pilot',
  air_date: 'December 2, 2013',
  episode: 'S01E01',
  url: 'https://rickandmortyapi.com/api/episode/1',
  created: '2017-11-10T12:56:33.798Z',
} as IEpisode;

const locationResponse = {
  dimension: '',
  name: '',
  residents: [''],
  type: '',
  url: '',
} as ILocationAll;

const mockedNavigationProp = {
  canGoBack: jest.fn().mockReturnValue(true),
  goBack: jest.fn(),
} as any;
const route = {
  key: '',
  name: 'Profile',
  params: {data: characterPageResponse.results[0]},
  path: '',
} as RouteProp<RootStackParamList, 'Profile'>;

const mockedNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({navigate: mockedNavigate}),
}));

jest.mock('../src/api');

it('Profile renders with chapter view', async () => {
  const mGetEpisodeByURL = jest.mocked(getEpisodeByURL);
  mGetEpisodeByURL.mockResolvedValue(episodePageResponse);

  const profile = render(
    <Profile navigation={mockedNavigationProp as any} route={route} />,
  );
  await waitFor(() => profile.getByTestId('ChapterRow.View'));
  expect(profile).toMatchSnapshot();
});

it('Location Row renders correctly without location url', async () => {
  const mGetLocationByURL = jest.mocked(getLocationByURL);
  mGetLocationByURL.mockResolvedValue({} as ILocationAll);
  const locationRow = render(
    <LocationRow location={{name: '', url: ''}} type="test" />,
  );
  const expandButton = locationRow.queryByTestId('PropRow.ExpandButton');
  expect(expandButton).toBeFalsy();
  expect(locationRow).toMatchSnapshot();
});

it('Location Row logs with incorrect location url', async () => {
  console.log = jest.fn();

  const mGetLocationByURL = jest.mocked(getLocationByURL);
  mGetLocationByURL.mockResolvedValue(null as any);

  const locationRow = render(
    <LocationRow location={{name: 'test', url: 'incorrect_url'}} type="test" />,
  );
  const expandButton = locationRow.queryByTestId('PropRow.ExpandButton');
  if (!expandButton) fail();
  fireEvent(expandButton, 'press');
  await waitFor(() => locationRow.queryByTestId('LocationExpandedCard.View'));
  expect(console.log).toHaveBeenCalledWith('Error fetching');
});

it('Location Row renders correctly with location url', async () => {
  const mGetLocationByURL = jest.mocked(getLocationByURL);
  mGetLocationByURL.mockResolvedValue(locationResponse);
  const locationRow = render(
    <LocationRow location={{name: 'test', url: 'sample_url'}} type="test" />,
  );
  const expandButton = locationRow.queryByTestId('PropRow.ExpandButton');
  expect(expandButton).toBeTruthy();
  if (!expandButton) fail();
  fireEvent(expandButton, 'press');
  await waitFor(() => locationRow.queryByTestId('LocationExpandedCard.View'));
  expect(locationRow.queryByTestId('LocationExpandedCard.View')).toBeTruthy();
  expect(locationRow).toMatchSnapshot();
});
