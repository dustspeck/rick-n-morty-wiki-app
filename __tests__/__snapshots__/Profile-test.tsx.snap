// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location Row renders correctly with location url 1`] = `
<View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View>
      <View
        style={
          {
            "flexDirection": "row",
            "marginVertical": 5,
            "paddingHorizontal": "10%",
          }
        }
      >
        <Text
          style={
            {
              "backgroundColor": "#E4DCCF",
              "borderRadius": 7,
              "color": "#476072",
              "fontSize": 13,
              "marginRight": 5,
              "minWidth": "20%",
              "paddingHorizontal": 10,
              "paddingVertical": 5,
              "textAlign": "center",
              "textAlignVertical": "center",
            }
          }
        >
          TEST
        </Text>
        <Text
          numberOfLines={2}
          style={
            [
              {
                "backgroundColor": "#E4DCCF55",
                "borderRadius": 7,
                "color": "#476072",
                "flex": 1,
                "fontSize": 15,
                "paddingHorizontal": 10,
                "paddingVertical": 5,
                "textAlign": "center",
              },
              {
                "borderBottomRightRadius": 0,
                "borderTopRightRadius": 0,
              },
            ]
          }
        >
          test
        </Text>
        <Text
          style={
            [
              {
                "backgroundColor": "#E4DCCF55",
                "borderRadius": 7,
                "color": "#476072",
                "flex": 1,
                "fontSize": 15,
                "paddingHorizontal": 10,
                "paddingVertical": 5,
                "textAlign": "center",
              },
              {
                "borderBottomLeftRadius": 0,
                "borderTopLeftRadius": 0,
                "maxWidth": 30,
                "padding": 0,
                "textAlignVertical": "center",
              },
            ]
          }
          testID="PropRow.ExpandButton"
        >
          ▲
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#E4DCCF55",
        "borderRadius": 14,
        "marginBottom": 20,
        "marginHorizontal": "10%",
        "overflow": "hidden",
        "padding": 14,
      }
    }
    testID="LocationExpandedCard.View"
  >
    <Text
      style={
        {
          "color": "#476072aa",
          "fontWeight": "bold",
        }
      }
    >
      Name  
      <Text
        style={
          {
            "color": "#476072",
            "fontWeight": "normal",
            "textDecorationLine": "underline",
          }
        }
      />
    </Text>
    <Text
      style={
        {
          "color": "#476072aa",
          "fontWeight": "bold",
        }
      }
    >
      Type  
      <Text
        style={
          {
            "color": "#476072",
            "fontWeight": "normal",
            "textDecorationLine": "underline",
          }
        }
      />
    </Text>
    <Text
      style={
        {
          "color": "#476072aa",
          "fontWeight": "bold",
        }
      }
    >
      Dimension  
      <Text
        style={
          {
            "color": "#476072",
            "fontWeight": "normal",
            "textDecorationLine": "underline",
          }
        }
      />
    </Text>
    <Text
      style={
        {
          "color": "#476072aa",
          "fontWeight": "bold",
        }
      }
    >
      Population  
      <Text
        style={
          {
            "color": "#476072",
            "fontWeight": "normal",
            "textDecorationLine": "underline",
          }
        }
      >
        1
      </Text>
    </Text>
  </View>
</View>
`;

exports[`Location Row renders correctly without location url 1`] = `
<View>
  <View>
    <View
      style={
        {
          "flexDirection": "row",
          "marginVertical": 5,
          "paddingHorizontal": "10%",
        }
      }
    >
      <Text
        style={
          {
            "backgroundColor": "#E4DCCF",
            "borderRadius": 7,
            "color": "#476072",
            "fontSize": 13,
            "marginRight": 5,
            "minWidth": "20%",
            "paddingHorizontal": 10,
            "paddingVertical": 5,
            "textAlign": "center",
            "textAlignVertical": "center",
          }
        }
      >
        TEST
      </Text>
      <Text
        numberOfLines={2}
        style={
          [
            {
              "backgroundColor": "#E4DCCF55",
              "borderRadius": 7,
              "color": "#476072",
              "flex": 1,
              "fontSize": 15,
              "paddingHorizontal": 10,
              "paddingVertical": 5,
              "textAlign": "center",
            },
            false,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Profile renders with chapter view 1`] = `
[
  <View
    style={
      {
        "backgroundColor": "#F0EBE3",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "height": 60,
          "paddingLeft": 5,
          "width": 60,
        }
      }
    >
      <Text
        style={
          {
            "color": "#2C3333",
            "fontSize": 45,
            "fontWeight": "bold",
            "textAlignVertical": "center",
          }
        }
      >
        ←
      </Text>
    </View>
  </View>,
  <View
    style={
      [
        {
          "backgroundColor": "#F0EBE3",
          "minHeight": 1334,
          "padding": 20,
        },
        undefined,
        {
          "paddingTop": 0,
        },
      ]
    }
  >
    <RCTScrollView
      centerContent={true}
      showsVerticalScrollIndicator={false}
      style={
        {
          "marginHorizontal": -20,
        }
      }
    >
      <View>
        <View>
          <View>
            <Image
              source={
                {
                  "uri": "https://rickandmortyapi.com/api/character/avatar/20.jpeg",
                }
              }
              style={
                [
                  {
                    "borderColor": "#476072",
                    "borderRadius": 375,
                    "borderWidth": 4,
                    "height": 375,
                    "marginBottom": -16,
                    "marginHorizontal": 187.5,
                    "marginTop": 10,
                    "width": 375,
                  },
                  {
                    "borderColor": "#476072",
                  },
                ]
              }
            />
            <View
              style={
                {
                  "alignItems": "center",
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  [
                    {
                      "borderRadius": 5,
                      "fontSize": 16,
                      "padding": 2,
                      "paddingHorizontal": 10,
                      "textAlign": "center",
                    },
                    {
                      "backgroundColor": "#476072",
                    },
                  ]
                }
              >
                UNKNOWN
              </Text>
            </View>
          </View>
          <Text
            numberOfLines={2}
            style={
              {
                "color": "#2C3333",
                "fontSize": 28,
                "fontWeight": "bold",
                "padding": 12,
                "textAlign": "center",
                "width": "100%",
              }
            }
          >
            Ants in my Eyes Johnson
          </Text>
          <View>
            <View>
              <Text
                style={
                  {
                    "color": "#476072",
                    "margin": 5,
                    "marginTop": 30,
                    "textAlign": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#47607233",
                    }
                  }
                >
                  ─────     
                </Text>
                PROPERTIES
                <Text
                  style={
                    {
                      "color": "#47607233",
                    }
                  }
                >
                       ─────
                </Text>
              </Text>
            </View>
            <View>
              <View>
                <View
                  style={
                    {
                      "flexDirection": "row",
                      "marginVertical": 5,
                      "paddingHorizontal": "10%",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "backgroundColor": "#E4DCCF",
                        "borderRadius": 7,
                        "color": "#476072",
                        "fontSize": 13,
                        "marginRight": 5,
                        "minWidth": "20%",
                        "paddingHorizontal": 10,
                        "paddingVertical": 5,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  >
                    GENDER
                  </Text>
                  <Text
                    numberOfLines={2}
                    style={
                      [
                        {
                          "backgroundColor": "#E4DCCF55",
                          "borderRadius": 7,
                          "color": "#476072",
                          "flex": 1,
                          "fontSize": 15,
                          "paddingHorizontal": 10,
                          "paddingVertical": 5,
                          "textAlign": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    Male
                  </Text>
                </View>
              </View>
            </View>
            <View>
              <View>
                <View
                  style={
                    {
                      "flexDirection": "row",
                      "marginVertical": 5,
                      "paddingHorizontal": "10%",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "backgroundColor": "#E4DCCF",
                        "borderRadius": 7,
                        "color": "#476072",
                        "fontSize": 13,
                        "marginRight": 5,
                        "minWidth": "20%",
                        "paddingHorizontal": 10,
                        "paddingVertical": 5,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  >
                    SPECIES
                  </Text>
                  <Text
                    numberOfLines={2}
                    style={
                      [
                        {
                          "backgroundColor": "#E4DCCF55",
                          "borderRadius": 7,
                          "color": "#476072",
                          "flex": 1,
                          "fontSize": 15,
                          "paddingHorizontal": 10,
                          "paddingVertical": 5,
                          "textAlign": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    Human
                  </Text>
                </View>
              </View>
            </View>
            <View>
              <View>
                <View
                  style={
                    {
                      "flexDirection": "row",
                      "marginVertical": 5,
                      "paddingHorizontal": "10%",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "backgroundColor": "#E4DCCF",
                        "borderRadius": 7,
                        "color": "#476072",
                        "fontSize": 13,
                        "marginRight": 5,
                        "minWidth": "20%",
                        "paddingHorizontal": 10,
                        "paddingVertical": 5,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  >
                    STATUS
                  </Text>
                  <Text
                    numberOfLines={2}
                    style={
                      [
                        {
                          "backgroundColor": "#E4DCCF55",
                          "borderRadius": 7,
                          "color": "#476072",
                          "flex": 1,
                          "fontSize": 15,
                          "paddingHorizontal": 10,
                          "paddingVertical": 5,
                          "textAlign": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    unknown
                  </Text>
                </View>
              </View>
            </View>
            <View>
              <View>
                <View
                  style={
                    {
                      "flexDirection": "row",
                      "marginVertical": 5,
                      "paddingHorizontal": "10%",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "backgroundColor": "#E4DCCF",
                        "borderRadius": 7,
                        "color": "#476072",
                        "fontSize": 13,
                        "marginRight": 5,
                        "minWidth": "20%",
                        "paddingHorizontal": 10,
                        "paddingVertical": 5,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  >
                    TYPE
                  </Text>
                  <Text
                    numberOfLines={2}
                    style={
                      [
                        {
                          "backgroundColor": "#E4DCCF55",
                          "borderRadius": 7,
                          "color": "#476072",
                          "flex": 1,
                          "fontSize": 15,
                          "paddingHorizontal": 10,
                          "paddingVertical": 5,
                          "textAlign": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    Human with ants in his eyes
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View>
            <View>
              <Text
                style={
                  {
                    "color": "#476072",
                    "margin": 5,
                    "marginTop": 30,
                    "textAlign": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#47607233",
                    }
                  }
                >
                  ─────     
                </Text>
                WHEREABOUTS
                <Text
                  style={
                    {
                      "color": "#47607233",
                    }
                  }
                >
                       ─────
                </Text>
              </Text>
            </View>
            <View>
              <View>
                <View
                  style={
                    {
                      "flexDirection": "row",
                      "marginVertical": 5,
                      "paddingHorizontal": "10%",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "backgroundColor": "#E4DCCF",
                        "borderRadius": 7,
                        "color": "#476072",
                        "fontSize": 13,
                        "marginRight": 5,
                        "minWidth": "20%",
                        "paddingHorizontal": 10,
                        "paddingVertical": 5,
                        "textAlign": "center",
                        "textAlignVertical": "center",
                      }
                    }
                  >
                    ORIGIN
                  </Text>
                  <Text
                    numberOfLines={2}
                    style={
                      [
                        {
                          "backgroundColor": "#E4DCCF55",
                          "borderRadius": 7,
                          "color": "#476072",
                          "flex": 1,
                          "fontSize": 15,
                          "paddingHorizontal": 10,
                          "paddingVertical": 5,
                          "textAlign": "center",
                        },
                        false,
                      ]
                    }
                  >
                    unknown
                  </Text>
                </View>
              </View>
            </View>
            <View>
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
              >
                <View>
                  <View
                    style={
                      {
                        "flexDirection": "row",
                        "marginVertical": 5,
                        "paddingHorizontal": "10%",
                      }
                    }
                  >
                    <Text
                      style={
                        {
                          "backgroundColor": "#E4DCCF",
                          "borderRadius": 7,
                          "color": "#476072",
                          "fontSize": 13,
                          "marginRight": 5,
                          "minWidth": "20%",
                          "paddingHorizontal": 10,
                          "paddingVertical": 5,
                          "textAlign": "center",
                          "textAlignVertical": "center",
                        }
                      }
                    >
                      LOCATION
                    </Text>
                    <Text
                      numberOfLines={2}
                      style={
                        [
                          {
                            "backgroundColor": "#E4DCCF55",
                            "borderRadius": 7,
                            "color": "#476072",
                            "flex": 1,
                            "fontSize": 15,
                            "paddingHorizontal": 10,
                            "paddingVertical": 5,
                            "textAlign": "center",
                          },
                          {
                            "borderBottomRightRadius": 0,
                            "borderTopRightRadius": 0,
                          },
                        ]
                      }
                    >
                      Interdimensional Cable
                    </Text>
                    <Text
                      style={
                        [
                          {
                            "backgroundColor": "#E4DCCF55",
                            "borderRadius": 7,
                            "color": "#476072",
                            "flex": 1,
                            "fontSize": 15,
                            "paddingHorizontal": 10,
                            "paddingVertical": 5,
                            "textAlign": "center",
                          },
                          {
                            "borderBottomLeftRadius": 0,
                            "borderTopLeftRadius": 0,
                            "maxWidth": 30,
                            "padding": 0,
                            "textAlignVertical": "center",
                          },
                        ]
                      }
                      testID="PropRow.ExpandButton"
                    >
                      ▼
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View>
            <View>
              <Text
                style={
                  {
                    "color": "#476072",
                    "margin": 5,
                    "marginTop": 30,
                    "textAlign": "center",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#47607233",
                    }
                  }
                >
                  ─────     
                </Text>
                FEATURED CHAPTERS
                <Text
                  style={
                    {
                      "color": "#47607233",
                    }
                  }
                >
                       ─────
                </Text>
              </Text>
            </View>
            <View
              style={
                {
                  "paddingTop": 20,
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": "#E4DCCF55",
                    "borderRadius": 4.666666666666667,
                    "marginHorizontal": "10%",
                    "marginVertical": 1,
                    "paddingHorizontal": "5%",
                    "paddingVertical": 8,
                    "textAlignVertical": "center",
                    "width": "80%",
                  }
                }
                testID="ChapterRow.View"
              >
                <View
                  style={
                    {
                      "flexDirection": "row",
                      "marginBottom": 4,
                    }
                  }
                >
                  <Text
                    ellipsizeMode="tail"
                    style={
                      {
                        "color": "#476072",
                        "flex": 1,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    S01E01
                  </Text>
                  <Text
                    ellipsizeMode="tail"
                    style={
                      {
                        "color": "#476072aa",
                        "fontSize": 13,
                      }
                    }
                  >
                    December 2, 2013
                  </Text>
                </View>
                <Text
                  ellipsizeMode="tail"
                  style={
                    {
                      "color": "#476072",
                    }
                  }
                >
                  Pilot
                </Text>
              </View>
            </View>
          </View>
          <View
            style={
              {
                "height": 200,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#47607255",
                  "fontSize": 14,
                  "marginTop": 25,
                  "textAlign": "center",
                }
              }
            >
              That's all we know!
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>,
]
`;
